### ISSUE TITLE
Double Key Entry Feature

### ISSUE THREAD

**Issue by mhabimana** [opencdms/Climsoft#216]
Double Key Entry Feature is important as it ensures that during the key-entry process, every data item from a paper form is entered twice by two different key entry operators. This assists in finding errors that arise from entering incorrect data while reading from a paper form. This feature is working well into v3.2 but it is not yet implemented into climsoft V4 under development.

**Comment by volloholic** [comment]
My understanding is that double data entry has currently been implemented as a process which updates data after it has been added to the observation initial table. 

I think it is worth a discussion about whether this is the desired approach. I can certainly see value in having a system which enables reentry of data which has already been accepted to the observation of data. However double data entry as implemented by the likes of CSPRo would normally be considered as a QC process which happens before data enters the equivalent of the observation initial table. 

I am not well placed to understand how double data entry is currently being used in Met services, but I do think there is potential benefit to using the implementation of data entry to have a conversation about how what may be useful going forward. To aid such a discussion I would like to clarify my language to ensure we understand one another.

By double data entry I understand a process of having two people digitising the same content to minimise errors. 

By reentry of data I understand a process of having a person digitising data for comparison with existing digitised content to correct potential errors.

So if I understand correctly then currently it is what I call reentry of data that has been implemented. Double data entry is therefor achieved by the first person entering the data and the second one using the reentry of data. This has implication particularly relating to other QC features. 

Depending on Met services needs and protocol this may be what is required. One of the reasons a dedicated double data entry system is desirable is that in some cases you do not want to consider data as entered until it has been double entered, in particular you would only run other QC on it once double data entry is completed. This would be consistent with the data entry tables being separate in the database as the currently are. 

An alternative which has been mentioned is to have the data entry going straight into an observation table, this would now be relatively easy to implement consistently through the user controls. Such a system would be consistent with the current implementation of double data entry as a reentry of the data because other QC procedures could then be implemented so that they could happen before the second data entry, after or both. 

I hope this is useful to start a larger discussion about how the data entry system as a whole can be adapted to include double data entry as an integral component.

**Comment by rdstern** [comment]
CLIMSOFT will increasingly be used by staff at individual Met stations, i.e. outside headquarters.  In some countries it may then be useful for them to send their data as entered, together with the paper records and for the 2nd entry to be by the data entry staff based at the headquarters.  I assume the proposed system for double entry ( or re-entry) will handle this situation.

There will then sometimes be the possibility that the data need to be used after the first entry, and before the checking.  One example is the 10-day agromet bulletins.  I wonder how this will work.  Sometimes this is considered apart from the main data entry - which is not desirable as it usually involves a separate, hence 3rd entry of the data, often into Excel, by the Ag Met section.

**Comment by rdstern** [comment]
That's great.  The old CLICOM had the problem that it was difficult to enter data into the database until a month was completed, and that made it very difficult for the 10-day bulletins.  So excellent that this problem no longer exists.

**Comment by volloholic** [comment]
This discussion seems useful already. From Rogers and Samuel's responses I believe we should be looking at what I put as the alternative implementation in my message above. If so then this does lead to a few changes I would propose in the current implementation. The most important relates to the database structure and QC processes.

Currently the data entry forms are tied to individual temporary tables in the database. These undergo some QC and then are uploaded to the observation initial table. Given the situation described by Roger above if the second data entry then happened at the headquarters which would now affect the observation initial table directly. Now either the QC needs to be duplicated in the observation initial table or this data would bypass some QC processes. 

Given this implementation there is no way of avoiding a duplication of QC procedures, which leads to possibilities for inconsistencies. This could be resolved by the data entry going directly to the observation initial table and there being a set of flags which indicate what QC it has undergone, whether it has been double entered and if it is considered as the 'current value'. Given such changes would involve slight changes to the data model we would need to ensure that they are thoroughly discussed and agreed upon before implementing.

**Comment by mhabimana** [comment]
@volloholic  You are right.Currently the data entry forms are tied to individual temporary tables in the database but the QC is being done at data entry level (e.g lower and upper limit checks),any value which is not trapped at entry level will be uploaded into the observationinitial where the QC is taking place. The upload on each and every  data entry forms will then transfer data from the data entry form table to observationinitial table and in this case transferred values will have QC status 0. In case the same value is transferred through the upload action,it will overwrite the previous one,duplicationis not allowed at this level.

**Comment by volloholic** [comment]
@mhabimana does this mean that the second data entry in the current implementation of the double data entry simply overwrites any values which are different?

**Comment by Steve-Palmer** [comment]
Sorry to be slow adding to this discussion. It might help to think about Processes and pathways for observations.
(1) The traditional process for observations is purely paper-based, and this is appropriate still for historical data held in paper archives, as well as some data still supplied by co-operating observers. Here the issue is to maximise that quantity of data keyed, while minimising the errors. This sort of key entry requires little or no meteorological knowledge or skill other than as a copy-typist. For this type of data entry, keying by two different people is a really good way of eliminating keying errors. Since the double-key-entry is only about ensuring that keying mistakes are resolved, then it is appropriate that this takes place before any of the keyed data reaches the main database tables. However, this process requires some of the station and instrument metadata to ensure that the keyed data belongs to the correct instrument and station, as well as basic limits checks on the keyed values.
For example, a lot of paper data was keyed in Sierra Leone, but it appears that one of the temporary staff employed to do this discovered that he could still get paid by copying the data keyed by others and changing the station name.... So now the whole database is suspect (as well as being on hard disks that no longer have a PC!). This type of double keying would have prevented this fraud.
(2) Much of the near-real-time data coming into Climsoft is keyed directly by the observer at the outstation. I would argue that there is no necessity to do double-key entry from the paper records which the observer writes at the same time as doing the observation. The observer should be using a specific key-entry form tailored to his/her station with appropriate QC checks built into the form, and my experience is that either professional or co-operating observers have a vested interest in doing a good job. If they do not, and produce an excessive number of QC queries at later stages in the QC process, then that should produce a management response, either retraining or disciplinary.
(3) There is room for debate about other pathways. For example, an office may act as a collecting centre for near-real-time observations by outstations or co-operating observers using phone. Here there is a possibility of transcription errors as well as typing errors. The originating observer may well supply a paper copy perhaps monthly. In this case, there certainly should be a QC process for comparing the data collected by phone with the paper records received later, but it is not certain that there is a necessity for these data to be re-keyed. Visual inspection by a trained QC operative may be enough - though if the skills are lacking, then re-keying may be cost-effective.
In summary, I am very keen that we use the QC status and QC flagging to record the stages of the QC process which have been completed, and what errors have been flagged.
